File: solutions/144/go/144.binarytreepreordertraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func preorderTraversal(root *TreeNode) []int {
    if root == nil {
		return []int{}
	}

	traversal := make([]int, 0)
	stack := make([]*TreeNode, 0)

	current := root
	for current != nil || len(stack) > 0 {
		for current != nil {
			traversal = append(traversal, current.Val)
			stack = append(stack, current)
			current = current.Left
		}
		current = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		current = current.Right
	}
	return traversal
}

File: solutions/144/js/144.binarytreepreordertraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
function preorderTraversal(root){
    const mValues = [];
    if (!root){
        return mValues;
    }
    
    traversal(mValues, root);
    
    return mValues;
}

function traversal(mValues, node) {
    mValues.push(node.val);
        
    if (node.left){
        traversal(mValues, node.left);
    }
    if (node.right){
        traversal(mValues, node.right);
    }
}